core/src/main/scala/chisel3/ElaboratedCircuit.scala

```diff
@@ -0,0 +1,111 @@
+  def name: String
+
+  /** The circuit as a string of FIRRTL IR
+    *
+    * This will include annotations automatically.
+    * For large circuits (> 2 GiB of text) use [[lazilySerialize]].
+    */
+  def serialize: String
+
+  /** The circuit as a string of FIRRTL IR
+    *
+    * For large circuits (> 2 GiB of text) use [[lazilySerialize]]
+    *
+    * @param annotations alternative annotations to include in the FIRRTL IR.
+    */
+  def serialize(annotations: Iterable[Annotation]): String
+
+  /** The circuit as a String of FIRRTL IR
+    *
+    * This will include annotations automatically.
+    * Serialized lazily to reduce peak memory use and support cicuits larger than 2 GiB.
+    */
+  def lazilySerialize: Iterable[String]
+
+  /** The circuit as a String of FIRRTL IR
+    *
+    * Serialized lazily to reduce peak memory use and support cicuits larger than 2 GiB.
+    *
+    * @param annotations alternative annotations to include in the FIRRTL IR.
+    */
+  def lazilySerialize(annotations: Iterable[Annotation]): Iterable[String]
+
+  /** The annotations created during elaboration of this circuit
+    *
+    * This does not include annotations passed to elaboration.
+    */
+  def annotations: Iterable[Annotation]
+
+  /** Definitions of all modules in this circuit */
+  def definitions: Iterable[Definition[BaseModule]]
+
+  /** The underlying circuit, for private use only */
+  private[chisel3] def _circuit: Circuit
+}
+
+private class ElaboratedCircuitImpl(circuit: Circuit, initialAnnotations: Seq[Annotation]) extends ElaboratedCircuit {
+
+  // Source locator needed for toDefinition
+  private implicit def sourceInfo: SourceInfo = UnlocatableSourceInfo
+
+  def name: String = circuit.name
+
+  def serialize: String = lazilySerialize.mkString
+
+  def serialize(annotations: Iterable[Annotation]): String = lazilySerialize(annotations).mkString
+
+  def lazilySerialize: Iterable[String] = {
+    val annotations = (initialAnnotations.view ++ circuit.firrtlAnnotations).flatMap {
+      case _: Unserializable     => None
+      case _: CustomFileEmission => None
+      case a => Some(a)
+    }.toVector
+    lazilySerialize(annotations)
+  }
+
+  def lazilySerialize(annotations: Iterable[Annotation]): Iterable[String] = {
+    val prelude = {
+      val dummyCircuit = circuit.copy(components = Nil)
+      val converted = Converter.convert(dummyCircuit)
+      val withAnnos = CircuitWithAnnos(converted, annotations.toVector)
+      Serializer.lazily(withAnnos)
+    }
+    val typeAliases: Seq[String] = circuit.typeAliases.map(_.name)
+    val modules = circuit.components.iterator.map(c => Converter.convert(c, typeAliases))
+    val moduleStrings = modules.flatMap { m =>
+      Serializer.lazily(m, 1) ++ Seq("\n\n")
+    }
+    prelude ++ moduleStrings
+  }
+
+  def annotations: Iterable[Annotation] = circuit.firrtlAnnotations
+
+  def definitions: Iterable[Definition[BaseModule]] = circuit.components.map(_.id.toDefinition)
```

Comment:
I'm changing this to just be `topDefinition` since that is sufficient for most purposes when combined with `Select`. 

It's likely we'll add a similar thing like `publicDefinitions` as well but I'll save that till we need it.